### ISSUE TITLE
Firefox Dynasty does not open `.webloc` files or links from other apps

### ISSUE THREAD

**Issue by Wowfunhappy** [i3roly/firefox-dynasty#26]
I'm using Firefox Dynasty 133.0.3 on OS X 10.9.5.

To reproduce:

1. Make sure Firefox Dynasty is set as your default browser.
2. Go to a website.
3. Drag the URL from your address bar onto your desktop. A `.webloc` file will be created.
4. Close the tab
5. Double click the `.webloc` file you created.

Firefox will come into focus, but it will not open a new website. Expected behavior is for it to re-open the website you were on previously (when you created the .webloc file).

As a workaround, you can drag and drop the webloc file into your tab bar, and it will open.

**Comment by Wowfunhappy** [comment]
...or maybe it's just all links from other apps that do this? I'm having trouble opening links from Apple Mail too. But it's strange because I could have sworn that was working this afternoon...

**Comment by RJVB** [comment]
What if you do `open http://www.foo.com` from a terminal?

The default browser setting combined with Apple LaunchServices and esp. when you have multiple versions of the same application installed the OS can get confused (quite literally) which to launch. And I've already seen Firefox quit without apparently doing anything, even when launched from the commandline.

**Comment by Wowfunhappy** [comment]
> What if you do `open http://www.foo.com`/ from a terminal?

The Firefox window comes into focus (or the app launches if it's closed), but foo.com does not open. I would think if this was a Launch Services issue, that wouldn't happen?

I assume you can't reproduce this?

**Comment by RJVB** [comment]
>The Firefox window comes into focus (or the app launches if it's closed), but foo.com does not open. I would think if this was a Launch Services issue, that wouldn't happen?
>
>I assume you can't reproduce this?

*I* can. In fact, I had already noticed it too.

It doesn't seem to be directly related to the way the open document events are handled by FF but rather how it handles documents that it's supposed to open itself.

When I use `open -a /path/to/the/chosen/Firefox.app <foo>` and `foo` is an image or text file, FF will come to and remain in the foreground but the file will open in Preview.app or TextEdit.app . Nothing happens it's an html file.

However, all those files open properly in a FF tab if I use the File/Open command. And that's weird because a priori and IIRC  a LaunchServices open event is supposed to correspond to using that File/Open menu item!

**Comment by Wowfunhappy** [comment]
> However, all those files open properly in a FF tab if I use the File/Open command. And that's weird because a priori and IIRC  a LaunchServices open event is supposed to correspond to using that File/Open menu item!

I'm pretty sure Firefox builds out its menu in a highly non-native way, there's the whole thing with [custom keyboard shortcuts not working](https://bugzilla.mozilla.org/show_bug.cgi?id=1333781) for example. So I wouldn't put too much stock in this.

**Comment by RJVB** [comment]
On Wednesday December 25 2024 16:51:17 Jonathan wrote:

>I'm pretty sure Firefox builds out its menu in a highly non-native way, there's the whole thing with [custom keyboard shortcuts not working](https://bugzilla.mozilla.org/show_bug.cgi?id=1333781) for example. So I wouldn't put too much stock in this.

Maybe. But items in the global menu bar still have to behave according to system rules, no?

Either way, the fact that FF comes to the front but then hands off documents to their designated viewer/editor application suggests that we're dealing with something downstream from the actual open document message handling.

I suppose we could write a proxy NSApp that collects arguments from the commandline as well as documents handed over via LaunchServices, and then uses the appropriate `execv()` function to invoke the bundle exec in the traditional Unix way (the process purposely overlaying itself with the actual browser application). Register that proxy as the default browser and we should be set.

**Comment by Wowfunhappy** [comment]
> Maybe. But items in the global menu bar still have to behave according to system rules, no?

...I wouldn't think so, no. I can create arbitrary menu items in the menu bar with whatever strings I want, does OS X check for certain strings and give them special powers?

> Register that proxy as the default browser and we should be set.

Wouldn't even need a proxy app, I can just swizzle Firefox's open url method... and then, since running running `/path/to/Firefox.app/Contents/MacOS/firefox https://foo.com` in Terminal seems to actually work, I guess I would just spawn an `sh` shell and do that.

But it would make me feel very stupid.

**Comment by RJVB** [comment]
>...I wouldn't think so, no. I can create arbitrary menu items in the menu bar with whatever strings I want, does OS X check for certain strings and give them special powers?

I do think that's exactly how it works, or at least used to work. Certain things hardly changed (for a long time) since the pre-OS X days. I can vaguely recall a translated MS Word version that wouldn't open documents in a similar fashion because the Open menu command had been mistranslated.

The thing here is of course that we have to assume that the mechanism does work in the official builds on supported OS versions. It would be really interesting to know if those official builds react the same way if you do e.g. `open -a /path/to/firefox.app foo.png` . For instance, opening a PNG via File/Open does NOT launch an external viewer for me; if official builds behave that way in reaction to `open` too than that could mean they handle that request somehow according to my `File/Open` hypothesis.

Another possibility is that something changed w.r.t. associated metadata - again, I only have a vague recollection here but IIRC Apple have been transitioning from the much more precise TYPE+CREATOR data they used in pre-OSX days via something associating a mime type to the file extension to one that uses only the file extension? But again I don't see how that can explain why FF doesn't open the documents it should be opening itself!

>Wouldn't even need a proxy app, I can just swizzle Firefox's open url method... and then, since running running `/path/to/Firefox.app/Contents/MacOS/firefox https://foo.com` in Terminal seems to actually work, I guess I would just spawn an `sh` shell and do that.

You'd probably want to dig out `argv[0]` (via NSArguments, IIRC) and use that to ensure you're using the right executable.

>
>But it would make me feel very stupid.

Esp. if it makes you end up in a recursive loop? ;)

**Comment by RJVB** [comment]
Plus, the advantage of a proxy app would be that you don't have to hack the actual browser app bundle but can update it freely, or even point the proxy to another app bundle. Heck, you could even allow renaming the proxy app bundle to be able to "drop an URL" onto a browser instance running a different profile, if you wanted ;)

**Comment by Wowfunhappy** [comment]
> Esp. if it makes you end up in a recursive loop? ;)

I'm assuming the shell command triggers a different code path (likely something UNIXy that doesn't interact with Objective-C), because otherwise why would one work and the other not work?

If I'm wrong, there are ways around that. I'm allergic to having multiple apps on my system that do the same thing, and I'm already hacking the app bundle to fix the aforementioned custom keyboard shortcuts Firefox bug (and I have a script that downloads new Firefox builds and re-applies the hack, I was smart and didn't write it in Applescript this time).

But hopefully we can find what's actually wrong and I don't have to do something stupid like this.

**Comment by i3roly** [comment]
fwiw this works fine on 10.13 and up so it's nothing i did, probably something i didn't do or account for.

too bad lol

**Comment by RJVB** [comment]
On Thursday December 26 2024 13:45:41 gagan sidhu wrote:
>fwiw this works fine on 10.13 and up so it's nothing i did, probably something i didn't do or account for.

So it doesn't work on 10.12 and below? That should make figuring out "what to do or account for" a bit quicker.

**Comment by Wowfunhappy** [comment]
Fwiw the open URLs handler is here: https://github.com/i3roly/firefox-dynasty/blob/a3d97e7693f6dd635ef268b217f23b5f2641b5b6/toolkit/xre/MacApplicationDelegate.mm#L360C1-L360C42

Despite what I said earlier I can't seem to swizzle this code, so I can't tell on my end whether this is actually getting called.

**Comment by i3roly** [comment]
i don’t think it’s that because when i tried to use 10.12, it still doesn’t work.

and if it works on 10.13, but not 10.12, there is nothing in those lines that suggests it’s where the problem is

Thanks,
Gagan

**Comment by Wowfunhappy** [comment]
I don't know what I'm doing, but:

Look at this commit: https://github.com/i3roly/firefox-dynasty/commit/ed28e7ee654135605be5d613327a98033dfd87a4#diff-57ca016379edd3f8b4a39dc89afe484548f0399b19672263020e0cf1c227ca09L351

In particular, look at how Firefox used to have a ton of code to register a handler (am I using the right terminology here?) for an Apple Event at startup:

```
  if (isGetURLEvent ||
      ([event eventClass] == 'WWW!' && [event eventID] == 'OURL')) {
    NSString* urlString =
        [[event paramDescriptorForKeyword:keyDirectObject] stringValue];
    NSURL* url = [NSURL URLWithString:urlString];
    [self openURL:url];
```

Chromium Legacy has similar code:

https://github.com/blueboxd/chromium-legacy/blob/51f5e07802b5f38866225ea0c42d13588e2d062c/chrome/browser/app_controller_mac.mm#L848
```
  // We need to register the handlers early to catch events fired on launch.
  NSAppleEventManager* em = [NSAppleEventManager sharedAppleEventManager];
  [em setEventHandler:self
          andSelector:@selector(getUrl:withReply:)
        forEventClass:kInternetEventClass
           andEventID:kAEGetURL];
  [em setEventHandler:self
          andSelector:@selector(getUrl:withReply:)
        forEventClass:'WWW!'    // A particularly ancient AppleEvent that dates
           andEventID:'OURL'];  // back to the Spyglass days.
```

But this code is not in upstream Chromium:
https://github.com/chromium/chromium/blob/e626cd16f6e67c55352630574942c74d25161456/chrome/browser/app_controller_mac.mm#L829

Presumably Bluebox changed it for a reason.

Feel free to tell me if I'm just seeing faces in clouds.

**Comment by RJVB** [comment]
On Thursday December 26 2024 16:42:35 Jonathan wrote:

>Oh, uh... this might be why!
>
>https://developer.apple.com/documentation/appkit/nsapplicationdelegate/application(_:open:)?language=objc
>
>It's 10.13+ only!

Yes, that could definitely be an explanation.

I'm not at my Mac ATM but I noticed something additional last night. Drag a document onto the FF dock icon and 1 of 2 things happens for me:
- the document is opened in the designated viewer/editor (e.g. with PDF and JPG files)
- a modal dialog is posted to warn that FF cannot open document X of type Y - this happens with PNG, TXT *and* HTML files.

Interestingly, it also mistakes PNGs for ICNS files....

What's strange here is that those rejected files are clearly registered as supported because I didn't have to force-open them. Plus of course the fact that this dialog doesn't appear when using `open -a /path/to/Firefox.app`.

Looking at https://searchfox.org/mozilla-central/source/toolkit/xre/MacApplicationDelegate.mm#268 I see that FF has the usual cross-platform application code for collecting all open-document events/messages at startup and then adding them to a standard commandline.

Anyway, @wowfunhappy, it seems that in order to swizzle this you could try to intercept all document-open messages in your swizzler and pass them to NSApplicationDelegate:openURLs? Individual calls with a 1-element NSArray<NSURL*> should be fine from the looks of it?

**Comment by RJVB** [comment]
René J.V. Bertin wrote on 20241227::15:22:13 re: "Re: [i3roly/firefox-dynasty] Firefox Dynasty does not open `.webloc` files or links from other apps (Issue #26)"
>Anyway, @wowfunhappy, it seems that in order to swizzle this you could try to intercept all document-open messages in your swizzler and pass them to NSApplicationDelegate:openURLs? Individual calls with a 1-element NSArray<NSURL*> should be fine from the looks of it?

This seems to be the commit that introduced the use of NSApplicationDelegate:openURLs :
https://searchfox.org/mozilla-central/diff/8653b8bcc8112e6d1568934ba53ed56ba56dea33/toolkit/xre/MacApplicationDelegate.mm#300

Can you swizzle the `- (BOOL)openURL:(NSURL*)url` and other methods/core that were/was removed by that commit? Reintroducing that could solve the issue and should be the equivalent of using a proxy application.

From the looks of it this could be reintroduced safely by adding a runtime check on OS 10.13 to the code that installs the event handler (and maybe on NOT OS 10.13 in 1 or 2 places).

I'm still curious though how certain document types would open in an external application while HTML documents would simply be ignored. And then I noticed something in the `openURLs` documentation:

```
urls

    An array of URLs to open. The list does not include URLs for which your app has a defined document type.
```

So HTML documents and the like would NOT be opened via `openURLs` ??!

**Comment by i3roly** [comment]
lol swizzle. you guys are so cute.

i dunno if this approach is stable though. the only downside is it seems to open the bookmark in a frehs tab alongside a blank tab.

i have to test and ensure 10.13+ is unaffected by your two's unbridled greed

https://github.com/user-attachments/assets/a2bc146c-d654-48d2-b28e-276bd1d9510e

**Comment by i3roly** [comment]
aight swizzle boys i'm gonna post an alpha 135a01 in a minute.

give a shot and mentally engorge yourselves as you so-often-do

**Comment by RJVB** [comment]
On Friday December 27 2024 10:35:58 gagan sidhu wrote:
>aight swizzle boys i'm gonna post an alpha 135a01 in a minute.

Not my term, just using it because it's faster to write haha.

135a already? Out of curiosity: any plans on the stable release version front?

**Comment by Wowfunhappy** [comment]
@i3roly Yep, the test build works! Thank you!!!

It does have the weird glitch you mentioned where if Firefox is not already running when you open a link, Firefox will launch with the link you opened in one tab, and also a blank tab. This isn't a big deal and highly preferable to links not working!

**Comment by RJVB** [comment]
Works here too and fully agreed with @wowfunhappy.

Using

```
> open -a /path/to/Firefox.app /other/path/to/doc.ext
```

FF opens the doc in a new tab in the current window, while dragging `dog.ext` onto the FF icon in the Dock opens it in a new window. Maybe that's normal behaviour?

A few terminal output observations that I hadn't seen before:

- `[GFX1-]: VideoBridgeParent receives IPC close with reason=AbnormalShutdown`
- `dyld: Registered code signature for /Volumes/VMs/FFDynasty-135a01/Firefox.app/Contents/MacOS/XUL` (usually 2-4 lines). Why would dyld all of a sudden start mentioning this?
- `*** Assertion failure in -[GeckoNSMenu insertItemWithTitle:action:keyEquivalent:atIndex:], /SourceCache/AppKit/AppKit-1265.21/Menus.subproj/NSMenu.m:687`
- `Mozilla has caught an Obj-C exception [NSInternalInconsistencyException: Invalid parameter not satisfying: (index >= 0) && (index <= [_itemArray count])]`

**Comment by RJVB** [comment]
And this, is this to be expected?

![Screen Shot 2024-12-27 at 23 34 19](https://github.com/user-attachments/assets/2aa00864-e617-4d89-9ac3-8da6ec012bff)

The bookmarks menu doesn't do this:
![Screen Shot 2024-12-27 at 23 35 41](https://github.com/user-attachments/assets/f8d834fa-8717-4f00-9686-a279bb97ed29)

**Comment by Wowfunhappy** [comment]
@RJVB I saw that too but assumed it was because this is a build from the main tree rather than stable.

**Comment by i3roly** [comment]
first one is definitely because it's central and not release. that commit hasn't been backed out yet, but it should be 

https://phabricator.services.mozilla.com/D232367

regarding the second one, i dunno.

**Comment by RJVB** [comment]
On Friday December 27 2024 14:40:11 gagan sidhu wrote:
>first one is definitely because it's central and not release. that commit hasn't been backed out yet, but it should be 

First one - my 1st screen shot, Second one the 2nd screenshot? That was just there for comparison, it looks as it should AFAICT.

Anything slated to be fixed is fine of course, the dyld messages could be because of some new alpha build setting that wasn't yet in 133a, but the assertion and the ObjC error seem like potential problems.

**Comment by i3roly** [comment]
follow the bugzilla to figure out which one.

i don’t see any assert in my minimal testing. it could be your illegal plugin triggering it too.

i had ublock as a test and i didn’t see any assert.

always try a clean profile before asking me about stuff like this, please.

**Comment by RJVB** [comment]
On Friday December 27 2024 15:27:40 gagan sidhu wrote:

Understood about the clean profile and I've also only seen the asserts once (I think while perusing this ticket), but

>it could be your illegal plugin triggering it too.

Erm, no, that's not very likely. These asserts have to do with menus and I am quite certain that I didn't use the unsigned extension. I've already pointed out that it may not have the "proper" signing but it is otherwise a fixed version of the "legal" plugin that can be installed in any Firefox build.

**Comment by i3roly** [comment]
plugins also have menus, and, as you saw in the video, there was no assert.

could just be a 10.9 issue. no idea

**Comment by RJVB** [comment]
Yes they have, but esp. when we're talking about addons/extensions they use just ask the host application to put them up when needed.

No point in flogging this until it becomes an actual, reproducible issue but did you ever happen to see whether FF implements each and every menu using NSMenu, including the ones hanging off the toolbar?

**Comment by Wowfunhappy** [comment]
> Just out of curiosity, did you ever happen to see whether FF implements each and every menu using NSMenu, including the ones hanging off the toolbar?

I can answer that: No. NSMenus are used for the secondary-click context menu and the global menu bar.

**Comment by RJVB** [comment]
On Tuesday December 31 2024 19:52:38 Jonathan wrote:
>I can answer that: No. NSMenus are used for the secondary-click context menu and the global menu bar.

That's what I expected.

**Comment by i3roly** [comment]
better bookmark fix will be rolled out in 1340. closing

**Comment by RJVB** [comment]
On Wednesday January 01 2025 06:16:37 gagan sidhu wrote:
>better bookmark fix will be rolled out in 1340. closing

Would that be the 134 release version?

Best wishes for 2025 to everyone here btw!

**Comment by i3roly** [comment]
yes, but that guy who posted about yosemite not working has delayed the push.

maybe i’ll still push but we’ll see.

Thanks,
Gagan

**Comment by i3roly** [comment]
meh it wasn't as good of a fix as i thought. i didn't realise it just opened the extra tab in a new window, which is worse actually. 

oh well. i don't have time to hunt down stuff like this. these guys just abuse certain calls at startup. sometimes it feels like the current team is destroying a strong codebase.

anyways i'm not here to have a discussion about it. just know that i am aware it's not as good as i thought it was and it'll probably go back to the behaviour you're already experiencing (extra tab)

**Comment by Wowfunhappy** [comment]
The extra tab really is fine, and I'm usually sensitive about these things! It doesn't even matter unless Firefox is closed at the time the link is clicked!

**Comment by RJVB** [comment]
I second that!

**Comment by i3roly** [comment]
try the latest build i just pushed. it should be fixed now.

maybe it breaks elsewhere. i expect WorkFromHome to do that and report back

Thanks,
Gagan

**Comment by Wowfunhappy** [comment]
Yes, there's no extra tab anymore!
